### ISSUE TITLE
Shorten "out of element" messageShortening "out of element" message

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#10446]
### Is your feature request related to a problem? Please describe.
In browse mode or documents, when leaving a particular element with the caret, "out of element" is announced.

### Describe the solution you'd like
Based on how this is implemented in braille, I'd like to suggest "element end". So, instead of "out of list", announce 'list end". This is somewhat shorter and therefore more intuitive. Furthermore, the change is relatively small.

### Describe alternatives you've considered
Suppress element ends if they are followed by an element of the same role, see #10420

**Comment by yozaam** [comment]
Wow, this is something I would have never thought of, I'll send a pull request right now, but I would like to ask the mods if there is a reason, to syntactically use the "out of" phrase, 
also I'm a little bit of a beginner to this codebase so forgive me if im going in the wrong direction,

[in the src/speech/intit file](https://github.com/nvaccess/nvda/blob/master/source/speech/__init__.py)
Also for eg. in the 		line 2107 & 1699	text=_("link") if link else _("out of %s")%_("link")

I have changed "out of %" to "%s end"

Also in other locations when we call _()
such as 
line 2143          text=_("out of grammar error")     
line 2174          _("out of table")

And other files like  [source/locale/th/LC_MESSAGES/nvda.po](https://github.com/nvaccess/nvda/blob/master/source/locale/th/LC_MESSAGES/nvda.po)    where we have msgid "out of %s" 
on 12 different lines including 2403

should I change these as well or are they relevant since it is better to say "out of grammar error" than "grammar error end" or those in nvda.po are not relevant to this issue

Would be very happy if you or somebody experienced could give me a tip, 
Am i going in the right direction?

**Comment by Adriani90** [comment]
You do not have to change the .po files. They will change automatically since they will get the right strings from the _() function out of the code in speechinit.py through Gettext when compiling.

**Comment by yozaam** [comment]
> You do not have to change the .po files. They will change automatically since they will get the right strings from the _() function out of the code in speechinit.py through Gettext when compiling.

Thank you, you completely solved my doubt,  so the two changes I have made are enough? And what about the other lines where we don't use %s like 
line 2143 text=("out of grammar error")
line 2174 _("out of table")

Should I change these as well,
Again, sorry I am new to this and I am very grateful for your guidance

**Comment by feerrenrut** [comment]
This is likely to be a controversial change, its very subjective and users are already familiar with the existing behavior. This also results in a slightly different meaning, "out of X" invokes images of a container, being surrounded by something, where "X end" is perhaps more linear.

I'm not trying to argue for one option over the other, but it could be a lot of work to convince people for a small change.

**Comment by feerrenrut** [comment]
Thinking this through further, “out of X” clearly describes a boundary transition, and is appropriate to announce between objects, where “X end” might be interpreted by some as the final element.

**Comment by Adriani90** [comment]
How about reporting with speech when entering a container and playing a sound when leaving it. In this case people would hear what king of container they are entering and would know when they leave it as soon as the sound plays. I guess this would be possible via speech refactor. Right?

**Comment by ruifontes** [comment]
Please, do not oblige all people to differenciate the sounds!


Rui Fontes


Às 17:06 de 15/04/2020, Adriani90 escreveu:

**Comment by Adriani90** [comment]
it is the sane sound for everything. it just indicates that you leave the entered element. The element would be reported by speech when entering it.

Von meinem iPhone gesendet

**Comment by lukaszgo1** [comment]
Sound announcement is not appropriate in this case as we are loosing a useful part of the message i.e what exactly we are leaving - important in case of nested containers. @feerrenrut While I understand your doubts about this change it brings two important advantages IMHO:
1. The changeable information is presented first which is always a good think
2. It is consistent with other screen readers - certainly with JAWS.

**Comment by Adriani90** [comment]
But you hear what you are entering, so it is clear what you are going to leave. For nested elements there could be the same sound when leaving the whole container, but a bit longer. 

Von meinem iPhone gesendet

**Comment by venkateshpotluri** [comment]
Hi, I read through the comments and understand that "<element> end" might result in users understanding that as <element> being the last element in the page. Would "<element> exited" be a lesser verbose message as opposed to "out of <element>"? It would make the experience less verbose while providing the same information. Of course, announcement for entry and sound for exit might be a feasible solution, but I am not sure on how this might effect braille-only users.

**Comment by feerrenrut** [comment]
@venkateshpotluri Yes, that is also a good suggestion. I propose we try the this as originally proposed by @leonardder, since that has the benefit that it also matches braille. We can see what our alpha users say, and if there is a backlash we can look at either using "blah exited", adding an option (yuck), or reverting the change.

**Comment by GdePaulo** [comment]
@leonardder Hello, we are four Msc Computer Science students from the Delft University of Technology currently following a course on software architecture. For this course, we are required to contribute and submit a pull request to an open source system. We decided to choose NVDA, as it seemed like an interesting project which can help a lot of people. As we do not have much experience contributing to open source systems, we thought that this looked like a good issue that we could start with. It seems that someone else already started working on this issue, but they haven't updated their pull request for over a year.

We were wondering, would it be okay if we took over this issue in order to try to solve it? And if so, @yozaam would you also be okay with this? Thank you!

**Comment by XLTechie** [comment]
I am not @leonardder or @feerrenrut, but I just wanted to welcome you to NVDA!

**Comment by GdePaulo** [comment]
Thank you for the welcome! This looks like a really nice project that I
hope we can contribute something positive to, even if it might not be the
biggest contribution.

**Comment by cary-rowen** [comment]
This question is as meaningful as #7232. Hope to see new progress.

**Comment by cary-rowen** [comment]
As an end user, I welcome you to join @GdePaulo And thank you

**Comment by GdePaulo** [comment]
> As an end user, I welcome you to join @GdePaulo And thank you

Hello, thank you for your reply! We have taken a look at the issue you have linked, and definitely understand the problem. There seems to be a lot of issues like the current one, and it would indeed make sense to allow this functionality to be configured, rather than incorporating these kinds of suggestions one by one every time they come up.

Unfortunately, tackling the proposed solution might be a bit out of our scope currently. As we do not have any prior experience with open source projects and NVDA, we were really hoping to, at least in the beginning, start with a simpler issue. Does this mean that there is no longer any interest in solving the current issue as it stands? If so, we might have to search for a different issue to work on. Thank you again for your response!

**Comment by jcsteh** [comment]
> Based on how this is implemented in braille, I'd like to suggest "element end". So, instead of "out of list", announce 'list end".

There's another, bigger difference between speech and braille, though. Braille explicitly renders "list end" at the end of the list. In contrast, speech outputs "out of list" when you transition out of a list. This distinction matters because if you're moving up instead of down and you move past the beginning of a list, you're not at the end of the list. You're at the start.

This is not to say that I necessarily disagree with this change. It would also mean that exit announcements would occur at the end of the last line (or whatever) of the element, rather than after you leave it. That might also alleviate some of the concerns around verbosity because excess verbosity tends to be more problematic at the start of an announcement rather than the end. However, I point this out because of this comment:

> Furthermore, the change is relatively small.

It's not as simple as just changing the announcement, nor will its impact on users be "just the words". This also needs to change what conditions we use to determine when to report element exits. Funnily enough, it probably means we can remove a bunch of caching code, since we'd no longer need to track elements we last entered. This is assuming that all relevant TextInfos support the `_startOfNode` and `_endOfNode` attributes, but if they don't, braille would be broken there.

**Comment by LeonarddeR** [comment]
I'm going to close this.

1. Several objections have been raised, of which the reply by @jcsteh in https://github.com/nvaccess/nvda/issues/10446#issuecomment-2134963346 has been well-grounded in examples
2. Even when we change this, it might result in objections raised by users later.
3. Additional suggestions have been made (e.g. to use sounds), but that is definitely out of scope for this issue.
